File: src/main/java/de/byjoker/myjfql/command/SelectCommand.java
package de.byjoker.myjfql.command;

import de.byjoker.myjfql.core.MyJFQL;
import de.byjoker.myjfql.database.*;
import de.byjoker.myjfql.lang.ColumnComparator;
import de.byjoker.myjfql.lang.ColumnFilter;
import de.byjoker.myjfql.lang.SortingOrder;
import de.byjoker.myjfql.server.session.Session;

import java.util.*;
import java.util.stream.Collectors;

@CommandHandler
public class SelectCommand extends Command {

    public SelectCommand() {
        super("select", Arrays.asList("COMMAND", "VALUE", "FROM", "WHERE", "LIMIT", "SORT", "ORDER", "PRIMARY-KEY"));
    }

    @Override
    public void handleCommand(CommandSender sender, Map<String, List<String>> args) {
        final Session session = sender.getSession();

        if (session == null) {
            sender.sendError("Session of this user is invalid!");
            return;
        }

        final Database database = session.getDatabase(MyJFQL.getInstance().getDatabaseService());

        if (database == null) {
            sender.sendError("No database is in use for this user!");
            return;
        }

        if (args.containsKey("VALUE")
                && args.containsKey("FROM")) {

            final String name = formatString(args.get("FROM"));

            if (name == null) {
                sender.sendError("Undefined name!");
                return;
            }

            if (!database.existsTable(name)) {
                sender.sendError("Table doesn't exists!");
                return;
            }

            final Table table = database.getTable(name);

            if (!sender.allowed(database.getId(), DatabaseAction.READ)) {
                sender.sendForbidden();
                return;
            }

            final List<String> values = new ArrayList<>();
            final Collection<String> tableStructure = table.getStructure();

            {
                List<String> strings = formatList(args.get("VALUE"));

                if (strings.size() == 0) {
                    sender.sendError("You need to specify the values!");
                    return;
                }

                strings = formatList(strings);

                if (!strings.contains("*")) {
                    for (String key : strings) {
                        if (!tableStructure.contains(key)) {
                            sender.sendError("Key doesn't exists!");
                            return;
                        }

                        values.add(key);
                    }
                } else {
                    values.addAll(table.getStructure());
                }
            }

            SortingOrder order = null;
            String sortedBy = null;
            int limit = -1;

            if (args.containsKey("LIMIT")) {
                try {
                    limit = formatInteger(args.get("LIMIT"));
                } catch (Exception ex) {
                    sender.sendError("Unknown limit!");
                    return;
                }

                if (limit <= 0) {
                    sender.sendError("Limit is too small!");
                    return;
                }
            }

            if (args.containsKey("SORT")) {
                final String sort = formatString(args.get("SORT"));

                if (sort == null) {
                    sender.sendError("Undefined sort item!");
                    return;
                }

                if (!tableStructure.contains(sort)) {
                    sender.sendError("Sort item doesn't exist!");
                }

                sortedBy = sort;
            }

            if (args.containsKey("ORDER")) {
                try {
                    order = SortingOrder.valueOf(formatString(args.get("ORDER")));
                } catch (Exception ex) {
                    sender.sendError("Unknown sort order!");
                    return;
                }

                if (sortedBy == null)
                    sortedBy = table.getPrimary();
            }

            if (args.containsKey("PRIMARY-KEY")) {
                final String primaryKey = formatString(args.get("PRIMARY-KEY"));

                if (primaryKey == null) {
                    sender.sendError("Undefined primary-key!");
                    return;
                }

                final Column column = table.getColumn(primaryKey);

                if (column == null) {
                    sender.sendError("Column was not found!");
                    return;
                }

                sender.sendResult(Collections.singletonList(column), values);
            } else if (args.containsKey("WHERE")) {
                List<Column> columns;

                try {
                    columns = ColumnFilter.filter(table, args.get("WHERE"), new ColumnComparator(sortedBy), order);
                } catch (Exception ex) {
                    sender.sendError(ex);
                    return;
                }

                if (columns == null) {
                    sender.sendError("Unknown statement error!");
                    return;
                }

                if (limit != -1) {
                    sender.sendResult(columns.stream().limit(limit).collect(Collectors.toList()), values);
                    return;
                }

                sender.sendResult(columns, values);
            } else {
                final Collection<Column> columns = sortedBy == null ? table.getColumns() : table.getColumns(new ColumnComparator(sortedBy), order);

                if (columns.size() == 0) {
                    sender.sendResult(new ArrayList<SimpleColumn>(), values);
                    return;
                }

                if (limit != -1) {
                    sender.sendResult(columns.stream().limit(limit).collect(Collectors.toList()), values);
                    return;
                }

                sender.sendResult(columns, values);
            }

            return;
        }

        sender.sendSyntax();
    }
}

File: src/main/java/de/byjoker/myjfql/command/SelectCommand.kt
package de.byjoker.myjfql.command

import de.byjoker.myjfql.core.MyJFQL
import de.byjoker.myjfql.database.Database
import de.byjoker.myjfql.database.DatabasePermissionLevel
import de.byjoker.myjfql.database.TableType
import de.byjoker.myjfql.lang.TableEntryComparator
import de.byjoker.myjfql.lang.TableEntryFilter
import de.byjoker.myjfql.network.session.Session
import de.byjoker.myjfql.util.Order
import de.byjoker.myjfql.util.ResultType
import org.jline.reader.ParsedLine
import java.util.stream.Collectors

@CommandHandler
class SelectCommand : Command("select", listOf("COMMAND", "VALUE", "FROM", "WHERE", "SORT", "ORDER", "LIMIT")) {

    override fun execute(sender: CommandSender, args: Map<String, List<String>>) {
        val databaseService = MyJFQL.getInstance().databaseService
        val session: Session? = sender.session

        if (session == null) {
            sender.sendError("Session of this user is invalid!")
            return
        }

        val database: Database? = session.getDatabase(databaseService)

        if (database == null) {
            sender.sendError("No database is in use for this user!")
            return
        }

        if (args.containsKey("VALUE") && args.containsKey("FROM")) {
            val from: String? = formatString(args["FROM"])

            if (from == null) {
                sender.sendError("Undefined table!")
                return
            }

            val table = database.getTable(from)

            if (table == null) {
                sender.sendError("Table doesn't exists!")
                return
            }

            if (!sender.allowed(database.id, DatabasePermissionLevel.READ)) {
                sender.sendForbidden()
                return
            }

            val structure = when {
                formatString(args["VALUE"]) == "*" -> table.structure
                else -> formatList(args["VALUE"]) ?: return
            }

            if (table.type == TableType.RELATIONAL && structure.stream()
                    .anyMatch { entry -> !table.structure.contains(entry) }
            ) {
                sender.sendError("Specified values don't match table structure!")
                return
            }

            var order: Order? = null
            var sortedBy: String? = null
            var limit = -1

            if (args.containsKey("LIMIT")) {
                limit = try {
                    formatInteger(args["LIMIT"])
                } catch (ex: Exception) {
                    sender.sendError("Unknown limit!")
                    return
                }

                if (limit <= 0) {
                    sender.sendError("Limit is too small!")
                    return
                }
            }

            if (args.containsKey("SORT")) {
                val sort = formatString(args["SORT"])

                if (sort == null) {
                    sender.sendError("Undefined sort item!")
                    return
                }

                if (!structure.contains(sort)) {
                    sender.sendError("Specified sort item doesn't match table structure!")
                    return
                }

                sortedBy = sort
            }

            if (args.containsKey("ORDER")) {
                order = try {
                    Order.valueOf(formatString(args["ORDER"])!!)
                } catch (ex: Exception) {
                    sender.sendError("Unknown sort order!")
                    return
                }

                if (sortedBy == null) sortedBy = table.primary
            }

            val resultType = if (table.type == TableType.RELATIONAL) ResultType.RELATIONAL else ResultType.DOCUMENT

            if (args.containsKey("PRIMARY-KEY")) {
                val primaryKey = formatString(args["PRIMARY-KEY"])

                if (primaryKey == null) {
                    sender.sendError("Undefined primary-key!")
                    return
                }

                val entry = table.getEntry(primaryKey)

                if (entry == null) {
                    sender.sendError("Entry was not found!")
                    return
                }

                sender.sendResult(listOf(entry), structure, resultType)
            } else if (args.containsKey("WHERE")) {
                val entries = try {
                    TableEntryFilter.filterByCommandLineArguments(
                        table, args["WHERE"], if (sortedBy == null) null else TableEntryComparator(
                            sortedBy
                        ), order
                    )
                } catch (ex: Exception) {
                    sender.sendError(ex)
                    return
                }

                if (entries == null) {
                    sender.sendError("Unknown statement error!")
                    return
                }

                if (limit != -1) {
                    sender.sendResult(
                        entries.stream().limit(limit.toLong()).collect(Collectors.toList()),
                        structure,
                        resultType
                    )
                    return
                }

                sender.sendResult(entries, structure, resultType)
            } else {
                val entries = if (sortedBy == null) table.entries else table.getEntries(
                    TableEntryComparator(sortedBy), order
                )

                if (entries.isEmpty()) {
                    sender.sendResult(listOf(), structure, resultType)
                    return
                }

                if (limit != -1) {
                    sender.sendResult(
                        entries.stream().limit(limit.toLong()).collect(Collectors.toList()),
                        structure,
                        resultType
                    )
                    return
                }

                sender.sendResult(entries, structure, resultType)
            }

            return
        }

        sender.sendSyntax()
    }

    override fun complete(sender: CommandSender, line: ParsedLine): List<String>? {
        sender.session ?: return null
        val database: Database = sender.session!!.getDatabase(MyJFQL.getInstance().databaseService) ?: return null

        if (!sender.allowed(database.id, DatabasePermissionLevel.READ)) {
            return null
        }

        val args = line.line().uppercase()
        val before = line.words()[line.wordIndex() - 1].uppercase()

        return when {
            !args.contains(" VALUE") -> listOf("value")
            before == "VALUE" -> listOf("*")
            !args.contains(" FROM") -> listOf("from")
            before == "FROM" -> database.tables.map { table -> table.name }.toList()
            !args.contains(" WHERE") && !args.contains(" PRIMARY-KEY") -> listOf("where", "primary-key")
            before == "WHERE" || before == "PRIMARY-KEY" -> null
            !args.contains(" SORT") && !args.contains(" ORDER") && !args.contains(" LIMIT") -> listOf(
                "sort",
                "order",
                "limit"
            )
            before == "SORT" -> null
            before == "ORDER" -> listOf("asc", "desc")
            before == "LIMIT" -> null
            else -> null
        }
    }

}
